File: soucre_code/from_On_Hand_3_android_wear/app/src/main/kotlin/vn/edu/uit/tpkd/wear/cogload/YawnClassifier.kt
package vn.edu.uit.tpkd.wear.cogload

import kotlin.math.abs

enum class YawnState {
    CALIBRATING,
    IDLE,
    MOUTH_OPEN,
    YAWNING,
    UNAVAILABLE,
}

data class YawnFrameObservation(
    val observedAtMonoMs: Long,
    val observedAtWallMs: Long,
    val jawOpen: Double?,
    val mouthAspectRatio: Double?,
)

data class YawnSeed(
    val totalCount: Int = 0,
    val alertCount: Int = 0,
    val totalDurationMs: Long = 0L,
    val recentEventTimesMs: List<Long> = emptyList(),
    val lastAlertAtMs: Long? = null,
)

data class YawnDetection(
    val state: YawnState,
    val observedAtMonoMs: Long,
    val jawOpen: Double? = null,
    val mouthAspectRatio: Double? = null,
    val calibrated: Boolean = false,
    val calibrationProgress: Int = 0,
    val calibrationRequired: Int = REQUIRED_SAMPLES,
    val calibrationSpanMs: Long = 0L,
    val totalCount: Int = 0,
    val alertCount: Int = 0,
    val totalDurationMs: Long = 0L,
    val eventsInWindow: Int = 0,
    val recentEventTimesMs: List<Long> = emptyList(),
    val lastAlertAtMs: Long? = null,
    val advisory: Boolean = false,
    val eventJustCounted: Boolean = false,
    val alertJustTriggered: Boolean = false,
    val persistenceChanged: Boolean = false,
    val reason: String = "",
) {
    companion object {
        const val REQUIRED_SAMPLES = 20
    }
}

class YawnClassifier(seed: YawnSeed = YawnSeed()) {
    private data class CalibrationSample(val monoMs: Long, val mar: Double, val jaw: Double)
    private data class Baseline(
        val center: Double,
        val noise: Double,
        val jawCenter: Double,
        val jawNoise: Double,
    )

    private val samples = ArrayDeque<CalibrationSample>()
    private var baseline: Baseline? = null
    private var candidateStartedAtMono: Long? = null
    private var lastOpenAtMono: Long? = null
    private var closedSinceMono: Long? = null
    private var peakJaw = 0.0
    private var peakMar = 0.0
    private var countedCurrentOpen = false
    private var currentYawnStartedAtMono: Long? = null
    private var lastEventAtMono = Long.MIN_VALUE
    private val recentEventTimesMs = ArrayDeque<Long>()
    private var totalCount = seed.totalCount.coerceAtLeast(0)
    private var alertCount = seed.alertCount.coerceAtLeast(0)
    private var totalDurationMs = seed.totalDurationMs.coerceAtLeast(0L)
    private var lastAlertAtMs = seed.lastAlertAtMs
    private var remoteEventsInWindow = 0
    private var lastRemoteSyncAtMs: Long? = null

    init {
        seed.recentEventTimesMs.sorted().forEach(recentEventTimesMs::addLast)
    }

    fun observe(observation: YawnFrameObservation): YawnDetection {
        pruneEvents(observation.observedAtWallMs)
        val jaw = observation.jawOpen
        val mar = observation.mouthAspectRatio
        if (jaw == null || mar == null || !jaw.isFinite() || !mar.isFinite()) {
            candidateStartedAtMono = null
            lastOpenAtMono = null
            peakJaw = 0.0
            peakMar = 0.0
            return snapshot(
                state = if (baseline == null) YawnState.CALIBRATING else YawnState.UNAVAILABLE,
                observation = observation,
                reason = "face_missing",
            )
        }

        if (baseline == null) return calibrate(observation, jaw, mar)
        val currentBaseline = requireNotNull(baseline)
        val marOpenThreshold = currentBaseline.center + maxOf(3.0 * currentBaseline.noise, 0.05)
        val marPeakThreshold = currentBaseline.center + maxOf(5.0 * currentBaseline.noise, 0.10)
        val marCloseThreshold = currentBaseline.center + maxOf(2.0 * currentBaseline.noise, 0.035)
        val jawOpenThreshold = maxOf(
            OPEN_JAW_FLOOR,
            currentBaseline.jawCenter + maxOf(4.0 * currentBaseline.jawNoise, 0.18),
        )
        val jawPeakThreshold = maxOf(
            PEAK_JAW_FLOOR,
            currentBaseline.jawCenter + maxOf(6.0 * currentBaseline.jawNoise, 0.28),
        )
        val jawCloseThreshold = currentBaseline.jawCenter + maxOf(2.0 * currentBaseline.jawNoise, 0.10)
        val open = jaw >= jawOpenThreshold && mar >= marOpenThreshold
        var eventJustCounted = false
        var alertJustTriggered = false
        var persistenceChanged = false
        var state = YawnState.IDLE

        if (open) {
            closedSinceMono = null
            val previousOpen = lastOpenAtMono
            if (previousOpen == null || observation.observedAtMonoMs - previousOpen > MAX_OPEN_SAMPLE_GAP_MS) {
                if (!countedCurrentOpen) candidateStartedAtMono = observation.observedAtMonoMs
                peakJaw = jaw
                peakMar = mar
            }
            lastOpenAtMono = observation.observedAtMonoMs
            peakJaw = maxOf(peakJaw, jaw)
            peakMar = maxOf(peakMar, mar)
            state = if (countedCurrentOpen) YawnState.YAWNING else YawnState.MOUTH_OPEN
            val candidateStart = candidateStartedAtMono
            if (!countedCurrentOpen && candidateStart != null &&
                observation.observedAtMonoMs - candidateStart >= OPEN_DURATION_MS &&
                (peakJaw >= jawPeakThreshold || peakMar >= marPeakThreshold) &&
                elapsedSince(observation.observedAtMonoMs, lastEventAtMono) >= EVENT_COOLDOWN_MS
            ) {
                countedCurrentOpen = true
                currentYawnStartedAtMono = candidateStart
                lastEventAtMono = observation.observedAtMonoMs
                totalCount += 1
                recentEventTimesMs.addLast(observation.observedAtWallMs)
                pruneEvents(observation.observedAtWallMs)
                eventJustCounted = true
                persistenceChanged = true
                state = YawnState.YAWNING
                val previousAlertAt = lastAlertAtMs
                if (eventsInWindowCount() >= ALERT_EVENT_COUNT &&
                    (previousAlertAt == null || observation.observedAtWallMs - previousAlertAt >= ALERT_COOLDOWN_MS)
                ) {
                    lastAlertAtMs = observation.observedAtWallMs
                    alertCount += 1
                    alertJustTriggered = true
                }
            }
        } else {
            if (jaw <= jawCloseThreshold || mar <= marCloseThreshold) {
                if (closedSinceMono == null) closedSinceMono = observation.observedAtMonoMs
            }
            val previousOpen = lastOpenAtMono
            if (!countedCurrentOpen && previousOpen != null &&
                observation.observedAtMonoMs - previousOpen > OPEN_GAP_MS
            ) {
                candidateStartedAtMono = null
                lastOpenAtMono = null
                peakJaw = 0.0
                peakMar = 0.0
            }
            val closedSince = closedSinceMono
            if (countedCurrentOpen && closedSince != null &&
                observation.observedAtMonoMs - closedSince >= CLOSE_HOLD_MS &&
                elapsedSince(observation.observedAtMonoMs, lastEventAtMono) >= EVENT_COOLDOWN_MS
            ) {
                currentYawnStartedAtMono?.let {
                    totalDurationMs += maxOf(OPEN_DURATION_MS, observation.observedAtMonoMs - it)
                    persistenceChanged = true
                }
                candidateStartedAtMono = null
                lastOpenAtMono = null
                peakJaw = 0.0
                peakMar = 0.0
                countedCurrentOpen = false
                currentYawnStartedAtMono = null
            }
            if (countedCurrentOpen) state = YawnState.YAWNING
        }

        return snapshot(
            state = state,
            observation = observation,
            jaw = jaw,
            mar = mar,
            reason = if (eventsInWindowCount() >= ALERT_EVENT_COUNT) "repeated_yawn" else "live",
            eventJustCounted = eventJustCounted,
            alertJustTriggered = alertJustTriggered,
            persistenceChanged = persistenceChanged,
        )
    }

    fun reset() {
        baseline = null
        samples.clear()
        candidateStartedAtMono = null
        lastOpenAtMono = null
        closedSinceMono = null
        peakJaw = 0.0
        peakMar = 0.0
        countedCurrentOpen = false
        currentYawnStartedAtMono = null
        lastEventAtMono = Long.MIN_VALUE
        remoteEventsInWindow = 0
        lastRemoteSyncAtMs = null
    }

    fun synchronizeRemote(
        remoteTotalCount: Int,
        remoteWindowCount: Int,
        observedAtMonoMs: Long,
        observedAtWallMs: Long,
    ): YawnDetection? {
        pruneEvents(observedAtWallMs)
        val boundedTotal = remoteTotalCount.coerceIn(0, 1_000_000)
        val boundedWindow = remoteWindowCount.coerceIn(0, 1_000)
        val totalChanged = boundedTotal > totalCount
        val windowChanged = boundedWindow != remoteEventsInWindow
        if (!totalChanged && !windowChanged) {
            lastRemoteSyncAtMs = observedAtWallMs
            return null
        }
        if (totalChanged) {
            totalCount = boundedTotal
            candidateStartedAtMono = null
            lastOpenAtMono = observedAtMonoMs
            closedSinceMono = null
            peakJaw = 0.0
            peakMar = 0.0
            countedCurrentOpen = true
            currentYawnStartedAtMono = null
            lastEventAtMono = observedAtMonoMs
        }
        remoteEventsInWindow = boundedWindow
        lastRemoteSyncAtMs = observedAtWallMs
        var alertJustTriggered = false
        val previousAlertAt = lastAlertAtMs
        if (eventsInWindowCount() >= ALERT_EVENT_COUNT &&
            (previousAlertAt == null || observedAtWallMs - previousAlertAt >= ALERT_COOLDOWN_MS)
        ) {
            lastAlertAtMs = observedAtWallMs
            alertCount += 1
            alertJustTriggered = true
        }
        return snapshot(
            state = if (totalChanged) YawnState.YAWNING else YawnState.IDLE,
            observation = YawnFrameObservation(observedAtMonoMs, observedAtWallMs, null, null),
            reason = if (eventsInWindowCount() >= ALERT_EVENT_COUNT) "repeated_yawn" else "remote_sync",
            eventJustCounted = totalChanged,
            alertJustTriggered = alertJustTriggered,
            persistenceChanged = totalChanged || windowChanged || alertJustTriggered,
        )
    }

    private fun calibrate(observation: YawnFrameObservation, jaw: Double, mar: Double): YawnDetection {
        var reason = if (jaw < CALIBRATION_JAW_MAX) "collecting_closed_mouth" else "mouth_moving"
        if (jaw < CALIBRATION_JAW_MAX) {
            samples.addLast(CalibrationSample(observation.observedAtMonoMs, mar, jaw))
        }
        while (samples.size > REQUIRED_SAMPLES) samples.removeFirst()
        var baselineReady = false
        if (samples.size == REQUIRED_SAMPLES && samples.last().monoMs - samples.first().monoMs >= CALIBRATION_SPAN_MS) {
            val values = samples.map(CalibrationSample::mar)
            val center = median(values)
            val noise = maxOf(0.002, median(values.map { abs(it - center) }))
            val jawValues = samples.map(CalibrationSample::jaw)
            val jawCenter = median(jawValues)
            val jawNoise = maxOf(0.002, median(jawValues.map { abs(it - jawCenter) }))
            if (noise <= MAX_CALIBRATION_MAD && jawNoise <= MAX_JAW_CALIBRATION_MAD) {
                baseline = Baseline(center, noise, jawCenter, jawNoise)
                samples.clear()
                baselineReady = true
                reason = "baseline_ready"
            } else {
                samples.removeFirst()
                reason = "mouth_moving"
            }
        }
        return snapshot(
            state = if (baselineReady) YawnState.IDLE else YawnState.CALIBRATING,
            observation = observation,
            jaw = jaw,
            mar = mar,
            reason = reason,
        )
    }

    private fun snapshot(
        state: YawnState,
        observation: YawnFrameObservation,
        jaw: Double? = null,
        mar: Double? = null,
        reason: String,
        eventJustCounted: Boolean = false,
        alertJustTriggered: Boolean = false,
        persistenceChanged: Boolean = false,
    ): YawnDetection {
        val span = if (samples.size > 1) samples.last().monoMs - samples.first().monoMs else 0L
        val windowCount = eventsInWindowCount()
        return YawnDetection(
            state = state,
            observedAtMonoMs = observation.observedAtMonoMs,
            jawOpen = jaw,
            mouthAspectRatio = mar,
            calibrated = baseline != null,
            calibrationProgress = if (baseline == null) samples.size else REQUIRED_SAMPLES,
            calibrationSpanMs = if (baseline == null) span else CALIBRATION_SPAN_MS,
            totalCount = totalCount,
            alertCount = alertCount,
            totalDurationMs = totalDurationMs,
            eventsInWindow = windowCount,
            recentEventTimesMs = recentEventTimesMs.toList(),
            lastAlertAtMs = lastAlertAtMs,
            advisory = windowCount >= ALERT_EVENT_COUNT,
            eventJustCounted = eventJustCounted,
            alertJustTriggered = alertJustTriggered,
            persistenceChanged = persistenceChanged,
            reason = reason,
        )
    }

    private fun pruneEvents(nowWallMs: Long) {
        while (recentEventTimesMs.isNotEmpty() && nowWallMs - recentEventTimesMs.first() > WINDOW_MS) {
            recentEventTimesMs.removeFirst()
        }
        val remoteAt = lastRemoteSyncAtMs
        if (remoteAt != null && nowWallMs - remoteAt > WINDOW_MS) remoteEventsInWindow = 0
    }

    private fun eventsInWindowCount(): Int = maxOf(recentEventTimesMs.size, remoteEventsInWindow)

    private fun median(values: List<Double>): Double {
        val sorted = values.sorted()
        val middle = sorted.size / 2
        return if (sorted.size % 2 == 1) sorted[middle] else (sorted[middle - 1] + sorted[middle]) / 2.0
    }

    private fun elapsedSince(now: Long, previous: Long): Long =
        if (previous == Long.MIN_VALUE) Long.MAX_VALUE else (now - previous).coerceAtLeast(0L)

    companion object {
        const val REQUIRED_SAMPLES = YawnDetection.REQUIRED_SAMPLES
        private const val CALIBRATION_SPAN_MS = 5_000L
        private const val MAX_CALIBRATION_MAD = 0.025
        private const val MAX_JAW_CALIBRATION_MAD = 0.04
        private const val CALIBRATION_JAW_MAX = 0.30
        private const val OPEN_JAW_FLOOR = 0.32
        private const val PEAK_JAW_FLOOR = 0.42
        private const val OPEN_DURATION_MS = 1_000L
        private const val OPEN_GAP_MS = 250L
        // Face runs at 2-2.5 FPS. This is the maximum gap between consecutive
        // open observations, not the permitted duration of a measured closure.
        private const val MAX_OPEN_SAMPLE_GAP_MS = 900L
        private const val CLOSE_HOLD_MS = 500L
        private const val EVENT_COOLDOWN_MS = 2_000L
        private const val WINDOW_MS = 10 * 60_000L
        private const val ALERT_COOLDOWN_MS = 10 * 60_000L
        private const val ALERT_EVENT_COUNT = 3
    }
}

object YawnRuntimeStore {
    @Volatile
    var snapshot: YawnDetection? = null
        private set

    @Synchronized
    fun update(value: YawnDetection) {
        snapshot = value
    }

    @Synchronized
    fun reset() {
        snapshot = null
    }
}

File: firmware/main/web/yawn_classifier.mjs
// SPDX-FileCopyrightText: 2026 vietdo1201
// SPDX-License-Identifier: Apache-2.0
export const YAWN_CLASSIFIER_VERSION = "YAWN_MAR_V4";

const REQUIRED_SAMPLES = 20;
const CALIBRATION_SPAN_MS = 5000;
const OPEN_JAW_FLOOR = 0.32;
const PEAK_JAW_FLOOR = 0.42;
const CALIBRATION_JAW_MAX = 0.30;
// The compact Web face task intentionally omits the blendshape head, so MAR
// must also be able to reject an open mouth during calibration by itself.
const CALIBRATION_MAR_MAX = 0.30;
const OPEN_DURATION_MS = 1000;
const OPEN_GAP_MS = 250;
// Face inference normally arrives every 350-500 ms. Consecutive open samples
// must not be mistaken for a >250 ms interruption merely because of cadence.
const MAX_OPEN_SAMPLE_GAP_MS = 900;
const CLOSE_HOLD_MS = 500;
const EVENT_COOLDOWN_MS = 2000;
const STALE_MS = 1000;
const WINDOW_MS = 10 * 60_000;
const ALERT_COOLDOWN_MS = 10 * 60_000;

const distance = (a, b) => Math.hypot(a.x - b.x, a.y - b.y);
const median = values => {
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[middle] : (sorted[middle - 1] + sorted[middle]) / 2;
};

function summarize(values) {
  const center = median(values);
  const noise = Math.max(0.002, median(values.map(value => Math.abs(value - center))));
  return {center, noise};
}

function validBaseline(value, fingerprint) {
  return value?.schema === 2 && value?.classifierVersion === YAWN_CLASSIFIER_VERSION &&
    value?.fingerprint === fingerprint && Number.isFinite(value.center) && value.center > 0 &&
    Number.isFinite(value.noise) && value.noise > 0 &&
    Number.isFinite(value.jawCenter) && value.jawCenter >= 0 &&
    Number.isFinite(value.jawNoise) && value.jawNoise > 0;
}

export function extractMouthAspectRatio(landmarks) {
  if (!Array.isArray(landmarks) || landmarks.length <= 308) return null;
  const upper = landmarks[13], lower = landmarks[14], left = landmarks[78], right = landmarks[308];
  if (![upper, lower, left, right].every(point => point && Number.isFinite(point.x) && Number.isFinite(point.y))) return null;
  const width = distance(left, right);
  if (!Number.isFinite(width) || width < 0.01) return null;
  const mar = distance(upper, lower) / width;
  return Number.isFinite(mar) && mar >= 0 && mar < 2 ? mar : null;
}

export class YawnClassifier {
  constructor({fingerprint, baseline = null} = {}) {
    this.fingerprint = fingerprint || "unknown";
    this.baseline = validBaseline(baseline, this.fingerprint) ? baseline : null;
    this.samples = [];
    this.resetTemporal();
    this.totalCount = 0;
    this.totalDurationMs = 0;
    this.events = [];
    this.lastAlertAt = -Infinity;
    this.lastResult = null;
  }

  resetTemporal() {
    this.candidateStartedAt = null;
    this.lastOpenAt = null;
    this.closedSince = null;
    this.peakJaw = 0;
    this.peakMar = 0;
    this.countedCurrentOpen = false;
    this.currentYawnStartedAt = null;
    this.lastEventAt = -Infinity;
  }

  reset() {
    this.baseline = null;
    this.samples = [];
    this.resetTemporal();
    this.lastResult = null;
  }

  persistedBaseline() {
    return this.baseline ? {...this.baseline} : null;
  }

  observe({sequence, timestampMs, landmarks, jawOpen}) {
    const mar = extractMouthAspectRatio(landmarks);
    // Number(null) is 0, which previously made an unavailable blendshape look
    // like a real closed-jaw measurement and permanently blocked Web yawns.
    const jaw = jawOpen == null ? null : Number(jawOpen);
    const hasJaw = Number.isFinite(jaw);
    if (!Number.isFinite(mar) || (jaw != null && !hasJaw)) {
      return this.unavailable(timestampMs, "face_missing");
    }

    if (!this.baseline) {
      const calibrationEligible = mar < CALIBRATION_MAR_MAX && (!hasJaw || jaw < CALIBRATION_JAW_MAX);
      if (calibrationEligible) {
        const previous = this.samples.at(-1);
        if (!previous || previous.sequence !== sequence) this.samples.push({sequence, timestampMs, mar, jaw});
      }
      while (this.samples.length > REQUIRED_SAMPLES) this.samples.shift();
      let reason = calibrationEligible ? "collecting_closed_mouth" : "mouth_moving";
      if (this.samples.length === REQUIRED_SAMPLES &&
          this.samples.at(-1).timestampMs - this.samples[0].timestampMs >= CALIBRATION_SPAN_MS) {
        const summary = summarize(this.samples.map(sample => sample.mar));
        const jawSamples = this.samples.map(sample => sample.jaw).filter(Number.isFinite);
        const jawSummary = jawSamples.length === REQUIRED_SAMPLES ? summarize(jawSamples) : {center: 0, noise: 0.002};
        if (summary.noise <= 0.025 && (!hasJaw || jawSummary.noise <= 0.04)) {
          this.baseline = {
            schema: 2,
            classifierVersion: YAWN_CLASSIFIER_VERSION,
            fingerprint: this.fingerprint,
            center: summary.center,
            noise: summary.noise,
            jawCenter: jawSummary.center,
            jawNoise: jawSummary.noise,
          };
          this.samples = [];
          reason = "baseline_ready";
        } else {
          this.samples.shift();
          reason = "mouth_moving";
        }
      }
      return this.publish({
        state: "CALIBRATING", timestampMs, jawOpen: jaw, mar,
        calibrationProgress: this.samples.length,
        calibrationRequired: REQUIRED_SAMPLES,
        calibrationSpanMs: this.samples.length > 1 ? this.samples.at(-1).timestampMs - this.samples[0].timestampMs : 0,
        reason,
        baselineChanged: reason === "baseline_ready",
      });
    }

    this.pruneEvents(timestampMs);
    const marThreshold = this.baseline.center + Math.max(3 * this.baseline.noise, 0.05);
    const marPeakThreshold = this.baseline.center + Math.max(5 * this.baseline.noise, 0.10);
    const closeThreshold = this.baseline.center + Math.max(2 * this.baseline.noise, 0.035);
    const jawThreshold = Math.max(OPEN_JAW_FLOOR,
      this.baseline.jawCenter + Math.max(4 * this.baseline.jawNoise, 0.18));
    const jawPeakThreshold = Math.max(PEAK_JAW_FLOOR,
      this.baseline.jawCenter + Math.max(6 * this.baseline.jawNoise, 0.28));
    const jawCloseThreshold = this.baseline.jawCenter + Math.max(2 * this.baseline.jawNoise, 0.10);
    // Web ships the landmarks-only face task to fit the fixed flash partition.
    // MAR is therefore the primary signal there; jaw remains a corroborating
    // signal on runtimes (such as Watch) that actually provide blendshapes.
    const open = mar >= marThreshold && (!hasJaw || jaw >= jawThreshold);
    let eventJustCounted = false;
    let alertJustTriggered = false;
    let persistenceChanged = false;
    let state = "IDLE";

    if (open) {
      this.closedSince = null;
      if (this.lastOpenAt == null || timestampMs - this.lastOpenAt > MAX_OPEN_SAMPLE_GAP_MS) {
        if (!this.countedCurrentOpen) this.candidateStartedAt = timestampMs;
        this.peakJaw = hasJaw ? jaw : 0;
        this.peakMar = mar;
      }
      this.lastOpenAt = timestampMs;
      if (hasJaw) this.peakJaw = Math.max(this.peakJaw, jaw);
      this.peakMar = Math.max(this.peakMar, mar);
      state = this.countedCurrentOpen ? "YAWNING" : "MOUTH_OPEN";
      if (!this.countedCurrentOpen && this.candidateStartedAt != null &&
          timestampMs - this.candidateStartedAt >= OPEN_DURATION_MS &&
          ((hasJaw && this.peakJaw >= jawPeakThreshold) || this.peakMar >= marPeakThreshold) &&
          timestampMs - this.lastEventAt >= EVENT_COOLDOWN_MS) {
        this.countedCurrentOpen = true;
        this.currentYawnStartedAt = this.candidateStartedAt;
        this.lastEventAt = timestampMs;
        this.totalCount += 1;
        this.events.push(timestampMs);
        eventJustCounted = true;
        persistenceChanged = true;
        state = "YAWNING";
        this.pruneEvents(timestampMs);
        if (this.events.length >= 3 && timestampMs - this.lastAlertAt >= ALERT_COOLDOWN_MS) {
          this.lastAlertAt = timestampMs;
          alertJustTriggered = true;
        }
      }
    } else {
      const clearlyClosed = mar <= closeThreshold || (hasJaw && jaw <= jawCloseThreshold);
      if (clearlyClosed) this.closedSince ??= timestampMs;
      if (!this.countedCurrentOpen && this.lastOpenAt != null && timestampMs - this.lastOpenAt > OPEN_GAP_MS) {
        this.candidateStartedAt = null;
        this.lastOpenAt = null;
        this.peakJaw = 0;
        this.peakMar = 0;
      }
      if (this.countedCurrentOpen && this.closedSince != null && timestampMs - this.closedSince >= CLOSE_HOLD_MS &&
          timestampMs - this.lastEventAt >= EVENT_COOLDOWN_MS) {
        if (this.currentYawnStartedAt != null) {
          this.totalDurationMs += Math.max(OPEN_DURATION_MS, timestampMs - this.currentYawnStartedAt);
          persistenceChanged = true;
        }
        this.candidateStartedAt = null;
        this.lastOpenAt = null;
        this.peakJaw = 0;
        this.peakMar = 0;
        this.countedCurrentOpen = false;
        this.currentYawnStartedAt = null;
      }
      if (this.countedCurrentOpen) state = "YAWNING";
    }

    const advisory = this.events.length >= 3;
    return this.publish({
      state, timestampMs, jawOpen: jaw, mar, marThreshold, jawThreshold,
      calibrationProgress: REQUIRED_SAMPLES, calibrationRequired: REQUIRED_SAMPLES,
      calibrationSpanMs: CALIBRATION_SPAN_MS, reason: advisory ? "repeated_yawn" : "live",
      totalCount: this.totalCount, totalDurationMs: this.totalDurationMs,
      eventsInWindow: this.events.length, advisory, eventJustCounted, alertJustTriggered,
      persistenceChanged,
    });
  }

  stale(timestampMs) {
    if (!this.lastResult || timestampMs - this.lastResult.timestampMs <= STALE_MS) return null;
    return this.unavailable(timestampMs, "stale");
  }

  unavailable(timestampMs, reason) {
    this.candidateStartedAt = null;
    this.lastOpenAt = null;
    this.peakJaw = 0;
    this.peakMar = 0;
    if (!this.countedCurrentOpen) this.closedSince = null;
    return this.publish({
      state: this.baseline ? "UNAVAILABLE" : "CALIBRATING", timestampMs,
      jawOpen: null, mar: null, calibrationProgress: this.samples.length,
      calibrationRequired: REQUIRED_SAMPLES, calibrationSpanMs: 0, reason,
    });
  }

  pruneEvents(timestampMs) {
    this.events = this.events.filter(value => timestampMs - value <= WINDOW_MS);
  }

  publish(value) {
    this.pruneEvents(value.timestampMs);
    this.lastResult = {
      totalCount: this.totalCount,
      totalDurationMs: this.totalDurationMs,
      eventsInWindow: this.events.length,
      advisory: this.events.length >= 3,
      baselineChanged: false,
      eventJustCounted: false,
      alertJustTriggered: false,
      persistenceChanged: false,
      ...value,
      calibrated: Boolean(this.baseline),
    };
    return this.lastResult;
  }
}
